archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,874 @@
+      linarith,
+      refine lt_add_of_pos_left 1 ((1 : ℕ).succ_mul_pos (succ_pos n)), },
+  end,
+  refine le_trans (bn_diff_le_geo_sum n) ((le_div_iff' h₁).mpr _),
+  rw mul_div (4 * (n.succ : ℝ) * (n.succ.succ : ℝ))
+    ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) (1 - (1 / (2 * (n.succ : ℝ) + 1)) ^ 2),
+  refine (div_le_one h₂).mpr _,
+  norm_num,
+  rw mul_div,
+  have h₃: 0 < (2 * ((n : ℝ) + 1) + 1) ^ 2 :=
+  begin
+    rw sq,
+    norm_cast,
+    simp only [canonically_ordered_comm_semiring.mul_pos,
+    succ_pos', and_self],
+  end,
+  refine (div_le_iff' h₃).mpr _,
+  rw mul_sub,
+  norm_num,
+  refine le_sub.mp _,
+  rw mul_one_div,
+  refine (div_le_iff' h₃).mpr ((le_mul_iff_one_le_right h₃).mpr
+    (le_sub_iff_add_le.mpr _)),
+  norm_cast,
+  rw sq,
+  linarith,
+end
+
+/-- For any `n`, we have `bn 1 - bn n ≤ 1/4` -/
+lemma bn_bounded_aux : ∀ (n : ℕ), bn 1 - bn n.succ ≤ 1 / 4 :=
+begin
+  let bn' : (ℕ → ℝ) := λ (k : ℕ), bn k.succ,
+  intro n,
+  calc
+  bn 1 - bn n.succ = bn' 0 - bn' n : rfl
+   ... = ∑ k in range n, (bn' k - bn' (k + 1))          : by rw ← (sum_range_sub' bn' n)
+   ... = ∑ k in range n, (bn k.succ - bn k.succ.succ)   : rfl
+   ... ≤ ∑ k in range n, 1 / (4 * k.succ * k.succ.succ) : sum_le_sum (λ k, λ hk, bn_sub_bn_succ k)
+   ... = ∑ k in range n, (1 / 4) * (1 / (k.succ * k.succ.succ)) :
+   begin
+     have hi : ∀ (k : ℕ), (1 : ℝ) / (4 * k.succ * k.succ.succ) =
+      (1 / 4) * (1 / (k.succ * k.succ.succ)) :=
+     begin
+       intro k,
+       norm_cast,
+       field_simp,
+       simp only [one_div, inv_inj],
+       ring_nf,
+     end,
+    refine sum_congr rfl (λ k, λ hk, hi k),
+   end
+   ... = 1 / 4 * ∑ k in range n, 1 / (k.succ * k.succ.succ) : by rw mul_sum
+   ... ≤ 1 / 4 * 1 :
+   begin
+     refine (mul_le_mul_left _).mpr _,
+     { exact div_pos one_pos four_pos },
+     { convert rat.cast_le.mpr (partial_sum_consecutive_reciprocals n),
+       rw rat_cast_sum,
+       push_cast,
+       exact rat.cast_one.symm },
+   end
+   ... = 1 / 4 : by rw mul_one,
+end
+
+/-- The sequence `bn` is bounded below by `3/4 - 1/2 * log 2`  for `n ≥ 1`. -/
+lemma bn_bounded_by_constant : ∀ (n : ℕ), 3 / (4 : ℝ) - 1 / 2 * log 2 ≤ bn n.succ :=
+begin
+  intro n,
+  calc
+  bn n.succ ≥ bn 1 - 1 / 4 : sub_le.mp (bn_bounded_aux n)
+   ... = (log ((1 : ℕ).factorial) - 1 / 2 * log (2 * (1 : ℕ)) - (1 : ℕ) *
+          log ((1 : ℕ) / (exp 1))) - 1 / 4 : by rw bn_formula 0
+   ... = 0 - 1 / 2 * log 2 - log (1 / (exp 1)) - 1 / 4 :
+      by simp only [factorial_one, cast_one, log_one, one_div, mul_one, log_inv, log_exp, mul_neg]
+   ... = -1 / 2 * log 2 - log (1 / (exp 1)) - 1 / 4 : by ring
+   ... = -1 / 2 * log 2 + 1 - 1 / 4  : by simp only [one_div, log_inv, log_exp, sub_neg_eq_add]
+   ... = -1 / 2 * log 2 + 3 / 4      : by ring
+   ... = 3 / (4 : ℝ) - 1 / 2 * log 2 : by ring,
+end
+
+/-- The sequence `bn` is bounded below. -/
+lemma bn_has_lower_bound : (lower_bounds (set.range (λ (k : ℕ), bn k.succ))).nonempty :=
```

Comment:
You can use `bdd_below`.

Reply:
When renaming and using `bdd_below`, I realized that this lemma is not even used, so I removed it.